stl/inc/random

```diff
@@ -1116,27 +1118,43 @@ protected:
     _Post_satisfies_(this->_Idx == 0)
 
         void _Refill_lower() { // compute values for the lower half of the history array
-        size_t _Ix;
-        for (_Ix = 0; _Ix < _Nx - _Mx; ++_Ix) { // fill in lower region
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _Mx];
-        }
+        if constexpr (_M_mod_n > 0) {
+            size_t _Ix;
+            for (_Ix = 0; _Ix < _Nx - _M_mod_n; ++_Ix) { // fill in lower region
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx + _M_mod_n];
+            }
+
+            for (; _Ix < _Nx - _One_mod_n; ++_Ix) { // fill in upper region (avoids modulus operation)
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _M_mod_n];
+            }
 
-        for (; _Ix < _Nx - 1; ++_Ix) { // fill in upper region (avoids modulus operation)
-            _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + 1] & _LMSK);
-            this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix - _Nx + _Mx];
+            if constexpr (_One_mod_n > 0) { // fill in _Ax[_Nx - 1]
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[0] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_M_mod_n - 1];
+            }
+        } else {
+            size_t _Ix;
+            for (_Ix = 0; _Ix < _Nx - _One_mod_n; ++_Ix) { // fill in values
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[_Ix + _Nx + _One_mod_n] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx];
+            }
+
+            if constexpr (_One_mod_n > 0) { // fill in _Ax[_Nx - 1]
+                _Ty _Tmp       = (this->_Ax[_Ix + _Nx] & _HMSK) | (this->_Ax[0] & _LMSK);
+                this->_Ax[_Ix] = (_Tmp >> 1) ^ (_Tmp & 1 ? _Px : 0) ^ this->_Ax[_Ix + _Nx];
+            }

```

Comment:
I am missing any relation to `_One_mod_n`

The other clause has `this->_Ax[_M_mod_n - 1]` at the end. Is this one here correct?

Reply:
`_Ix` is `_Nx - 1` here, thus
- `_Ix + _Nx` = `_Ix + _Nx + _M_mod_n` = `(_Ix + _Nx + _M_mod_n) % (_Nx * 2)` in this clause,
- `_M_mod_n - 1` = `_Ix - _Nx + _M_mod_n` = `(_Ix + _Nx + _M_mod_n) % (_Nx * 2)` in the other clause,

so both are correct.

I should probably substitute `_Ix` for `_Nx - 1` in both clauses.